### PR TITLE
FETtec OneWire on top of AP_ESC_Telem backend v8

### PR DESCRIPTION
**Pull request by amilcarlucas**
This one is based on top of a newer version #16725 with https://github.com/andyp1per/ardupilot/pull/19 on top of it.

requires #16725

It is mostly similar to #16850 , #16963, #17012, #17063, #17201, #17225 and #17236 but it is based on top of some other commits.

- [x] feature complete:
  - [x] control motor speed
  - [x] copy ESC telemetry data into MAVLink telemetry
  - [x] save ESC telemetry data in Dataflash logs
  - [x] use RPM telemetry for dynamic notch filter frequencies
  - [x] sum the current telemetry info from all ESCs and use it as virtual battery current monitor sensor
  - [x] average the voltage telemetry info and use it as virtual battery voltage monitor sensor
  - [x] average the temperature telemetry info and use it as virtual battery temperature monitor sensor
  - [x] report telemetry communication error rate in the dataflsh logs
- [x] passes CI
- [x] compiles when HAL_WITH_ESC_TELEM = 0
- [x] compiles when HAL_WITH_ESC_TELEM = 1
- [x] compiles when  HAL_AP_FETTECONEWIRE_ENABLED = 0
- [x] compiles when  HAL_AP_FETTECONEWIRE_ENABLED = 1
- [x] compiles when  HAL_AP_FETTECONEWIRE_ENABLED = 1 and HAL_WITH_ESC_TELEM = 0
- [x] protocol is documented in the `README.md` file in the same location as the `.cpp` implementation file
- [x] [configuration instructions](https://github.com/amilcarlucas/ardupilot_wiki/blob/patch-11/common/source/docs/common-fettec-onewire.rst) are on PR ArduPilot/ardupilot_wiki#3335 the for the [ardupilot wiki](https://ardupilot.org/ardupilot/)
- [x] flight tested with a Hex frame using the first 6 channels (FETtec ID 1 to 6)
- [x] mavlink telemetry works (analyzed a `.TLOG` file to see that [ESC_TELEMETRY_1_TO_4](https://mavlink.io/en/messages/ardupilotmega.html#ESC_TELEMETRY_1_TO_4) and [ESC_TELEMETRY_5_TO_8](https://mavlink.io/en/messages/ardupilotmega.html#ESC_TELEMETRY_5_TO_8) messages are present - using MissionPlanner->Ctrl-F->MavLinkInspector->Vehicle1->Comp1 - and their contents correct)
- [x] `.bin` dataflash log files have ESC telemetry data in ESC1..6
- [x] four independent [notch filters](https://ardupilot.org/copter/docs/common-imu-notch-filtering.html) work (used a hex for testing, the last two motors are ignored) 
- INS_HNTCH_ENABLE = 1,
- [INS_HNTCH_MODE](https://ardupilot.org/copter/docs/parameters.html#ins-hntch-mode-harmonic-notch-filter-dynamic-frequency-tracking-mode) = 3 (ESC telemetry),
- [INS_HNTCH_OPTS](https://ardupilot.org/copter/docs/parameters.html#ins-hntch-opts-harmonic-notch-filter-options) = 2 (dynamic notch)
- [INS_HNTCH_REF](https://ardupilot.org/copter/docs/parameters.html#ins-hntch-ref) = 1
- [INS_HNTCH_FREQ](https://ardupilot.org/copter/docs/parameters.html#ins-hntch-freq) = 60
- [INS_HNTCH_BW](https://ardupilot.org/copter/docs/parameters.html#ins-hntch-bw) = 30
- [x] [IMU noise levels](https://ardupilot.org/copter/docs/common-imu-fft.html) measured before [notch filtering](https://ardupilot.org/copter/docs/common-imu-notch-filtering.html)
- [INS_LOG_BAT_MASK](https://ardupilot.org/copter/docs/parameters.html#ins-log-bat-mask) = 1
- [INS_LOG_BAT_OPT](https://ardupilot.org/copter/docs/parameters.html#ins-log-bat-opt-batch-logging-options-mask) = 0
- @teezeefpv please add a  [MissionPlanner FFT picture](https://ardupilot.org/planner/docs/common-measuring-vibration.html) here like described in the [noise analysis page](https://ardupilot.org/copter/docs/common-imu-batchsampling.html)
- [x] [IMU noise levels](https://ardupilot.org/copter/docs/common-imu-fft.html) measured after [notch filtering](https://ardupilot.org/copter/docs/common-imu-notch-filtering.html)
- [INS_LOG_BAT_OPT](https://ardupilot.org/copter/docs/parameters.html#ins-log-bat-opt-batch-logging-options-mask) = 2
- @teezeefpv please add a [MissionPlanner FFT picture](https://ardupilot.org/planner/docs/common-measuring-vibration.html) here like described in the [noise analysis page](https://ardupilot.org/copter/docs/common-imu-batchsampling.html)
- [x] Telemetry measurements replace the need for power monitoring hardware when [BATT_MONITOR](https://ardupilot.org/copter/docs/parameters.html#batt-monitor-battery-monitoring) = 9 (ESC). Used a normal battery monitor on [BATTERY2](https://mavlink.io/en/messages/ardupilotmega.html#BATTERY2) and compared the MAVLink telemetry results with [BATTERY_STATUS](https://mavlink.io/en/messages/common.html#BATTERY_STATUS) using `MissionPlanner`->`Ctrl-F`->`MavLinkInspector`->`Vehicle1`->`Comp1`:
  - [x] voltage
  - [x] current
  - [x] consumed power (will not match, but is present and valid on BATTERY_STATUS when id == 0)
  - [x] temperature (will not match, but is present and valid on BATTERY_STATUS when id == 0)
- [x] Make sure that BATT1 data (from FETtec ESCs) and BATT2 data (from a normal battery monitor) match in the `.bin` dataflash log files:
  - [x] voltage
  - [x] current
  - [ ] consumed power
  - [x] temperature (will not match, but is present and valid on BATT1)
- [x] use ArduPilot's coding guidelines and naming conventions

Compared to master this PR needs just 592 bytes for ArduCopter on a Durandal flight controller :)

I also [backported most of this to ArduCopter-4.0.7](https://github.com/ArduPilot/ardupilot/pull/16874)

Flight logs will be uploaded to [Copting GmbH Google Drive](https://drive.google.com/drive/folders/15L3EWuRJ4yhAwteYoU15bQeZkAslj1bL?usp=sharing). It also contains older logs from previous firmware versions. Tested on a Hexacopter with Air gear 450 motors, FETtec OneWire ESCs and CubeOrange.

### CODE REVIEW

**Comment by khancyr** [comment]
Could you just rebase and force push instead of creating 10PR per week for the same thing ?

**Comment by amilcarlucas** [comment]
@khancyr I am working with two companies on this, and this is the easiest way to document progress and compare versions. If it is not desired I can move this to my private repo, but then it will be harder to get it reviewed.
Lots of commits from these versions have now flown into #16725. So far this form of colaboartion has been very beneficial.

**Comment by teezeefpv** [comment]
image
Compiled and flashed V8. I just tested it in my desk, but I see, that the temperature is not correct. Current consumed is switching between ESC consumed and analog sensor consumed. 
Btw. 35227 are 35Ah? That would be to much.

Temperature is still wrong. I removed *100 in AP_FETtecOneWire.cpp for testing. While the  ESC telemetry temp values were 0, the battery_status began switching between something around 2400 and 32. 
32 would be the correct temp and 2400/10 could be the correct temp for the internal sensor. 
I think there is something confusion with centi-degree...

**Comment by amilcarlucas** [comment]
@teezeefpv Thanks for testing, I will fix that tomorrow afternoon.

**Comment by amilcarlucas** [comment]
@teezeefpv the fixes from v8 were: 2a59fb9 , a68c25d and 6c99ca0.
Do not change AP_FETtecOneWire.cpp. The bugs are on libraries/AP_BattMonitor/AP_BattMonitor_ESC.cpp

**Comment by amilcarlucas** [comment]
@andyp1per created PR andyp1per/ardupilot#22 with further fixes. I guess what is missing now is moving the dataflash logging to the AP_ESC_Telem class.

**Comment by teezeefpv** [comment]
:-) Thank you, looks very promising. 

When it's ID 1 the temp looks good (if it is correct, that it is multiplied with 100). 
image

I just noticed, that the motor RPM show 114 when the motors stand still.

**Comment by andyp1per** [comment]
@teezeefpv is this after the motors have been running? Do the ESCs send telemetry even when they are stopped?

**Comment by teezeefpv** [comment]
From Copting Google Drive -> "Arducopter_V8_07bac889" directory:
Log "00000125.BIN"
INS_LOG_BAT_OPT = 0
~5 Minutes flight time
image

Log "00000126.BIN"
INS_LOG_BAT_OPT = 2
~12 Minutes flight time
image

Log "00000127.BIN"
INS_LOG_BAT_OPT = 2
~3 Minutes flight time
image


The voltage is looking good to me. The "spike" looks like normal breaking. No unusual spikes this time.
image

Current:
image

Battery temp from ESC telem:
image

Consumption generally looks ok, except for the spike. 
image

The spikes seems to be random different values. I tried to find the spikes in Mission Planner, but they do no appear there as far as I can see. Maybe they aren't captured with the lower sample rate.
image

In addition... are there scaling errors with analog voltage/current sensors?
image

**Comment by amilcarlucas** [comment]
@teezeefpv please select the same y axis number when comparing the values. It's on the top left side of the plot.ardupilot.org gui.
Looks like the consumed current is a factor 1000 too big. :( Otherwise from that all test pass!!! :)

**Comment by amilcarlucas** [comment]
@teezeefpv take a look at #17216 
And it would be nice to add something similar to #17092 to this driver

File: libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp [hunk 1]
```diff
@@ -539,25 +541,52 @@ uint8_t AP_FETtecOneWire::init_escs()
...
 }
 
+/**
+    calculates crc tx error rate for incoming packages. It converts the CRC error counts into percentage
+    @param esc_id id of ESC, that the error is calculated for
+    @param esc_error_count the error count given by the esc
+    @param increment_only if this is set to 1 it only increases the message count and returns 0. If set to 1 it does not increment but gives back the error count.
+    @return the error in percent
+*/
+float AP_FETtecOneWire::calc_tx_crc_error_perc(uint8_t esc_id, uint16_t esc_error_count, uint8_t increment_only){
```
Comment (amilcarlucas) [inline libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:551]: Nope, the idea we discussed on the phone is to have this code in the ESC itself, it should transmit the error_rate*10000 over the wire as a uint16, and the ardupilot just calls:
'''
update_rpm(_telem_req_type-1, requested_telemetry[3]*100*2/pole_count.get(), requested_telemetry[5]);
'''

File: libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp [hunk 2]
```diff
@@ -611,14 +640,14 @@ int8_t AP_FETtecOneWire::check_for_tlm(uint16_t* Telemetry)
...
+    @param telemetry 16bit array where the read telemetry will be stored in.
     @param motorCount the count of motors that should get values send
     @param tlm_request the requested telemetry type (telem_type::XXXXX)
     @return the telemetry request if telemetry was available, -1 if dont
 */
-int8_t AP_FETtecOneWire::escs_set_values(uint16_t* motor_values, uint16_t* Telemetry, uint8_t motorCount,
+int8_t AP_FETtecOneWire::escs_set_values(uint16_t* motor_values, uint16_t* telemetry, uint8_t motorCount,
         uint8_t tlm_request)
-{
+{
```
Comment (amilcarlucas) [inline libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp:650]: whitespace

**Comment by amilcarlucas** [comment]
Closed in favor of #17363

**Pull request closed without merging**